### PR TITLE
Feature/issue 13- HomeScreen - Update UI

### LINKED ISSUE #13: [Feature] HomeScreen/ Update UI
### Overview
- //  update Ui

### Impact
- // Home Screen

### Screenshots (Optional)
- // Capture affected screens
Simulator Screen Shot - iPhone 12 Pro - 2021-10-14 at 11 42 55
Simulator Screen Shot - iPhone 12 Pro - 2021-10-14 at 14 40 56
Simulator Screen Shot - iPhone 12 Pro - 2021-10-14 at 14 41 02
Simulator Screen Shot - iPhone 12 Pro - 2021-10-14 at 14 41 08

### PR DESCRIPTION
**Pull request by minh5896**
## Prerequisite
- [x] Self-review :monocle_face:
- [x] No Crashes ☠️
---

### Root cause
- //[Feature] HomeScreen/ Update UI #13

### Solution
- // use MVVM

### Impact
- // Home Screen

### Screenshots (Optional)
- // Capture affected screen
Simulator Screen Shot - iPhone 12 Pro - 2021-10-14 at 11 42 55
Simulator Screen Shot - iPhone 12 Pro - 2021-10-14 at 14 41 08

Simulator Screen Shot - iPhone 12 Pro - 2021-10-14 at 14 40 56
Simulator Screen Shot - iPhone 12 Pro - 2021-10-14 at 14 41 02

### CODE REVIEW

File: MovieApp/SharedData/GlobalConstants.swift [hunk 1]
```diff
@@ -14,6 +14,22 @@ enum GlobalConstants {
...
+	static let continueMovieCell = "ContinueCell"
+	static let listMovieCell = "ListMovieCell"
+	static let tvShowMovieCell = "TvShowViewCell"
+	static let detailViewController = "DetailsViewController"
+}
+
+enum ConstantsCell {
+	static let previewCollectionCell = "PreviewCollectionViewCell"
+	static let continueItemCell = "ContinueItemCell"
+	static let listItemCellCV = "ListItemCellCV"
```
Comment (namnh92) [inline MovieApp/SharedData/GlobalConstants.swift]: @minh5896 : Move to Model folder of Home screen

File: MovieApp/SharedData/GlobalConstants.swift [hunk 2]
```diff
@@ -23,3 +39,31 @@ enum Icon {
...
+		case .europe:
+			return "Europe"
+		case .romance:
+			return "Romance"
+		case .action:
+			return "Action"
+
+		}
+	}
+}
```
Comment (namnh92) [inline MovieApp/SharedData/GlobalConstants.swift]: @minh5896 : Move to Model folder of Home screen

File: MovieApp/Screens/Home/Views/PreviewCell/PreviewCollectionViewCell/PreviewCollectionViewCell.swift [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+//
+//  Created by MinhDev on 10/12/21.
+//
+
+import UIKit
+
+class PreviewCollectionViewCell: UICollectionViewCell {
+	// MARK: - Outlets
+	@IBOutlet weak var imagePreview: UIImageView!
+	// MARK: - LifeCycle
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/PreviewCell/PreviewCollectionViewCell/PreviewCollectionViewCell.swift]: @minh5896 : Add line space

File: MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift [hunk 4]
```diff
@@ -7,10 +7,65 @@
...
 
+protocol clickedCell: AnyObject {
+	func goDetailView(celldata: MovieViewModel)
+}
+
 class PreviewCell: UITableViewCell {
+	// MARK: - Iboutlets
+	@IBOutlet weak var previewCollectionView: UICollectionView!
+	// MARK: - Variables
+	weak var delegate: clickedCell?
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift]:
```suggestion
	// MARK: - Iboutlets
	@IBOutlet weak var previewCollectionView: UICollectionView!
	
	// MARK: - Variables
	weak var delegate: clickedCell?
```

File: MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift [hunk 5]
```diff
@@ -7,10 +7,65 @@
...
+	var previewData = [MovieViewModel]() {
+		didSet {
+			previewCollectionView.reloadData()
+		}
+	}
 	// MARK: - Life cycles
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift]:
```suggestion
	}
	
	// MARK: - Life cycles
```

File: MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift [hunk 6]
```diff
@@ -7,10 +7,65 @@
...
+		previewCollectionView.dataSource = self
+		previewCollectionView.register(UINib(nibName: ConstantsCell.previewCollectionCell, bundle: Bundle.main), forCellWithReuseIdentifier: ConstantsCell.previewCollectionCell)
+		self.backgroundColor = .black
+	}
+	// MARK: - Life cycles
+	func configure(preview: [MovieViewModel]) {
+		previewData = preview
+	}
+}
+// MARK: - UICollectionViewDataSource
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift]:
```suggestion
	}
	
	// MARK: - Life cycles
	func configure(preview: [MovieViewModel]) {
		previewData = preview
	}
}

// MARK: - UICollectionViewDataSource
```

File: MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift [hunk 7]
```diff
@@ -6,14 +6,56 @@
...
+		collectionListView.dataSource = self
+		collectionListView.register(UINib(nibName: ConstantsCell.listItemCellCV, bundle: Bundle.main), forCellWithReuseIdentifier: ConstantsCell.listItemCellCV)
+		
+	}
+	// MARK: - configure
+	func configure(with listMovie: [MovieViewModel]) {
+		self.listData = listMovie
+	}
+}
+	// MARK: - UICollectionViewDataSource
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift]:
```suggestion
// MARK: - UICollectionViewDataSource
```

File: MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift [hunk 8]
```diff
@@ -6,14 +6,56 @@
...
+	}
+
+	func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+		guard let cell = self.collectionListView.dequeueReusableCell(withReuseIdentifier: ConstantsCell.listItemCellCV, for: indexPath) as? ListItemCellCV else { return UICollectionViewCell() }
+		cell.configure(with: listData[indexPath.row])
+		return cell
+	}
+	
+}
+	// MARK: - UICollectionViewDelegateFlowLayout
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift]:
```suggestion
// MARK: - UICollectionViewDelegateFlowLayout
```

File: MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift [hunk 9]
```diff
@@ -7,10 +7,65 @@
 
 import UIKit
 
+protocol clickedCell: AnyObject {
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift]:
```suggestion
protocol PreviewCellDelegate: AnyObject {
```

File: MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCellCV.swift [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+//  ListItemCellCV.swift
+//  MovieApp
+//
+//  Created by MinhDev on 10/12/21.
+//
+
+import UIKit
+import Model
+
+class ListItemCellCV: UICollectionViewCell {
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCellCV.swift]:
```suggestion
class ListItemCell: UICollectionViewCell {
```

@minh5896 : What is CV ? Please define name follow with suffix *Cell*

File: MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCellCV.swift [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+	// MARK: - IbOutlets
+	@IBOutlet private weak var imageListItem: UIImageView!
+	// MARK: - Life cycles
+	override func awakeFromNib() {
+		super.awakeFromNib()
+		imageListItem.image = UIImage(named: "image 2")
+		
+	}
+	// MARK: - configure
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCellCV.swift]:
```suggestion
	// MARK: - IbOutlets
	@IBOutlet private weak var imageListItem: UIImageView!
	
	// MARK: - Life cycles
	override func awakeFromNib() {
		super.awakeFromNib()
		imageListItem.image = UIImage(named: "image 2")
		
	}
	
	// MARK: - configure
```

File: MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCellCV.swift [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+import UIKit
+import Model
+
+class ListItemCellCV: UICollectionViewCell {
+	// MARK: - IbOutlets
+	@IBOutlet private weak var imageListItem: UIImageView!
+	// MARK: - Life cycles
+	override func awakeFromNib() {
+		super.awakeFromNib()
+		imageListItem.image = UIImage(named: "image 2")
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCellCV.swift]:
```suggestion
		imageListItem.image = UIImage(named: "icon_xxx")
```
@minh5896 Please define image named follow prefix *icon_*

File: MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift [hunk 13]
```diff
@@ -6,14 +6,56 @@
...
 	// MARK: - Life cycles
 	override func awakeFromNib() {
 		super.awakeFromNib()
-		// Initialization code
+		collectionListView.delegate = self
+		collectionListView.dataSource = self
+		collectionListView.register(UINib(nibName: ConstantsCell.listItemCellCV, bundle: Bundle.main), forCellWithReuseIdentifier: ConstantsCell.listItemCellCV)
+		
+	}
+	// MARK: - configure
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift]:
```suggestion
	}
	
	// MARK: - configure
```

File: MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift [hunk 14]
```diff
@@ -6,14 +6,56 @@
...
 
 import UIKit
+import Model
 
 class ListMovieCell: UITableViewCell {
 	// MARK: - Outlets
+	@IBOutlet private weak var collectionListView: UICollectionView!
+	// MARK: - Variables
+	weak var delegate: clickedCell?
+	var listData = [MovieViewModel]() {
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift]: @minh5896 : Please make it private set.

File: MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift [hunk 15]
```diff
@@ -7,10 +7,65 @@
...
 	override func awakeFromNib() {
 		super.awakeFromNib()
-		// Initialization code
+		previewCollectionView.delegate = self
+		previewCollectionView.dataSource = self
+		previewCollectionView.register(UINib(nibName: ConstantsCell.previewCollectionCell, bundle: Bundle.main), forCellWithReuseIdentifier: ConstantsCell.previewCollectionCell)
+		self.backgroundColor = .black
+	}
+	// MARK: - Life cycles
+	func configure(preview: [MovieViewModel]) {
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift]:
```suggestion
	func configure(preview: [IMovieViewModel]) {
```

@minh5896 : Don't forget SOLID, replace by interface

**Review by namnh92 — CHANGES_REQUESTED** [review]

File: MovieApp/SharedData/GlobalConstants.swift (hunk 1 above)
Comment (minh5896) [reply MovieApp/SharedData/GlobalConstants.swift]: fix complete

File: MovieApp/SharedData/GlobalConstants.swift (hunk 2 above)
Comment (minh5896) [reply MovieApp/SharedData/GlobalConstants.swift]: fix complete

File: MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift (hunk 4 above)
Comment (minh5896) [reply MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift]: fix complete

File: MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift (hunk 5 above)
Comment (minh5896) [reply MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift]: fix complete

File: MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift (hunk 6 above)
Comment (minh5896) [reply MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift]: fix complete

File: MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift (hunk 7 above)
Comment (minh5896) [reply MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift]: fix complete

File: MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift (hunk 8 above)
Comment (minh5896) [reply MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift]: fix complete

File: MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift (hunk 9 above)
Comment (minh5896) [reply MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift]: fix complete

File: MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCellCV.swift (hunk 10 above)
Comment (minh5896) [reply MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCellCV.swift]: fix complete

File: MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCellCV.swift (hunk 11 above)
Comment (minh5896) [reply MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCellCV.swift]: fix complete

File: MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCellCV.swift (hunk 12 above)
Comment (minh5896) [reply MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCellCV.swift]: fix complete

File: MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift (hunk 13 above)
Comment (minh5896) [reply MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift]: fix complete

File: MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift (hunk 14 above)
Comment (minh5896) [reply MovieApp/Screens/Home/Views/ListCell/ListMovieCell.swift]: fix complete

File: MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift (hunk 15 above)
Comment (minh5896) [reply MovieApp/Screens/Home/Views/PreviewCell/PreviewCell.swift]: fix complete

File: MovieApp/Screens/Home/Views/HomeMovieCell/HomeMovieCell.swift [hunk 16]
```diff
@@ -18,21 +19,17 @@ class HomeMovieCell: UITableViewCell {
...
-	// MARK: - Variables
-	private var viewModel: IMovieViewModel!
-	// MARK: - Life cycles
+	@IBOutlet private weak var imageHomeMovieCell: UIImageView!
+
+// MARK: - Life cycles
 	override func awakeFromNib() {
 		super.awakeFromNib()
-		self.backgroundView = UIImageView(image: UIImage(named: "image2"))
 	}
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/HomeMovieCell/HomeMovieCell.swift:27]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Code4Fun-Group/Movie-App/blob/6d8da324e91a4d6141b0bbfcfe90b541f3760d2d/MovieApp/Screens/Home/Views/HomeMovieCell/HomeMovieCell.swift#L27"></span>Override methods which only call super can be removed</td>
    </tr>
  </tbody>
</table>

File: MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCell.swift [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+import Model
+
+class ListItemCell: UICollectionViewCell {
+// MARK: - IbOutlets
+	@IBOutlet private weak var imageListItem: UIImageView!
+
+// MARK: - Life cycles
+	override func awakeFromNib() {
+		super.awakeFromNib()
+	}
```
Comment (namnh92) [inline MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCell.swift:18]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/Code4Fun-Group/Movie-App/blob/6d8da324e91a4d6141b0bbfcfe90b541f3760d2d/MovieApp/Screens/Home/Views/ListCell/ListItemCell/ListItemCell.swift#L18"></span>Override methods which only call super can be removed</td>
    </tr>
  </tbody>
</table>

**Comment by namnh92** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/Code4Fun-Group/Movie-App/blob/6d8da324e91a4d6141b0bbfcfe90b541f3760d2d/MovieApp/Screens/Details/DetailsViewController.swift#L14">MovieApp/Screens/Details/DetailsViewController.swift#L14</a> - Override methods which only call super can be removed</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by namnh92 — APPROVED** [review]

**Pull request merged**